plugin/pulpcore/plugin/download/asyncio/base.py

```diff
@@ -0,0 +1,153 @@
+        else:
+            fd, self.path = tempfile.mkstemp(dir=os.getcwd())
+            self._writer = os.fdopen(fd, mode='wb')
+        self.expected_digests = expected_digests
+        self.expected_size = expected_size
+        self._digests = {n: hashlib.new(n) for n in Artifact.DIGEST_FIELDS}
+        self._size = 0
+
+    def handle_data(self, data):
+        """
+        Write data to the file object and compute its digests.
+
+        All subclassed downloaders are expected to pass all data downloaded to this method. Similar
+        to the hashlib docstring, repeated calls are equivalent to a single call with
+        the concatenation of all the arguments: m.handle_data(a); m.handle_data(b) is equivalent to
+        m.handle_data(a+b).
+
+        Args:
+            data (bytes): The data to be handled by the downloader.
+        """
+        self._writer.write(data)
+        self._record_size_and_digests_for_data(data)
+
+    def _record_size_and_digests_for_data(self, data):
+        """
+        Record the size and digest for an available chunk of data.
+
+        Args:
+            data (bytes): The data to have its size and digest values recorded.
+        """
+        for algorithm in self._digests.values():
+            algorithm.update(data)
+        self._size = self._size + len(data)
+
+    @property
+    def artifact_attributes(self):
+        """
+        A property that returns a dictionary with size and digest information. The keys of this
+        dictionary correspond with :class:`pulpcore.plugin.models.Artifact` fields.
+        """
+        attributes = {'size': self._size}
+        for algorithm in Artifact.DIGEST_FIELDS:
+            attributes[algorithm] = self._digests[algorithm].hexdigest()
+        return attributes
+
+    def validate_digests(self):
+        """
+        Validate all digests validate if ``expected_digests`` is set
+
+        Raises:
+            :class:`~pulpcore.plugin.download.asyncio.DigestValidationError`: When any of the
+                ``expected_digest`` values don't match the digest of the data passed to
+                :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.handle_data`.
+        """
+        if self.expected_digests:
+            for algorithm, expected_digest in self.expected_digests.items():
+                if expected_digest != self._digests[algorithm].hexdigest():
+                    raise DigestValidationError()
+
+    def validate_size(self):
+        """
+        Validate the size if ``expected_size`` is set
+
+        Raises:
+            :class:`~pulpcore.plugin.download.asyncio.SizeValidationError`: When the
+                ``expected_size`` value doesn't match the size of the data passed to
+                :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.handle_data`.
+        """
+        if self.expected_size:
+            if self._size != self.expected_size:
+                raise SizeValidationError()
+
+    async def run(self):
```

Comment:
Assuming this returns a `coroutine` and does not actually *run* the download, I suggest another name like `get_coroutine()` would be more accurate.

Reply:
In looking at other docs (like aiohttp which has lots of coroutines) the method names of co-routines are the method names you would expect with or without the co-routine pattern. This method runs the downloader so I want to keep the name `run`. With that reasoning I'm going to leave as-is.

Interestingly as Sphinx improves so will our docs once this is merged and released:  https://github.com/sphinx-doc/sphinx/pull/1826 I am adding a line to each `run()` method so that until this Sphinx improvement is merged it will still be clear it's a co-routine in the docs. In the code it should be totally clear with the `async` keyword in the method definition already.